solveset consolidation and related changes

smichr opened the pull request:
closes #10555 
closes #10579 

@MSeeker , it also bothered me that the code for solveset_real and solveset_complex was _so similar_ and yet existed as two functions. As to the reason for the expansion code, it is attempting to get the equations into a standardized form and not (as the inline note said) oscillating between related forms. It could be that the complex functions didn't attempt the same sort of simplification as the real form so that's why it appeared only in the real form. Anyway, it's a lot more unified now with a less destructive expansion being used.

It was interesting to see a failing test pass after reconstruction.

I also discovered a very nasty sort of simple mistake (`foo is x` instead of `foo == x` that was causing different code paths to be taken). That didn't show up until after reconstruction: sometimes tests would pass and sometimes they would fail. This is what necessitated the "minimal simplification" that I added: when the poly-solver was handling the equations (instead of the inverter) the simplification was being done; after the changes the poly-solver was being detected as unnecessary so that special simplification was added in an unobtrusive place.

This does not yet resolve issue #10430 

``` python
>>> from sympy import *
>>> x = var('x'); eq = (x-1)*sin(x)

>>> solveset(eq, x)
ConditionSet(x, Eq((x - 1)*sin(x), 0), Complexes((-oo, oo) x (-oo, oo), False))

>>> solveset(sin(x), x)
ImageSet(Lambda(_n, 2*_n*pi), Integers()) U ImageSet(Lambda(_n, 2*_n*pi + pi), Integers())
```

smichr committed:
consolidate solveset code

- solveset passes off to _solveset once all preparation has finished
- _solveset sets the functions and the pertinent domains to use
- solveset_real is now a wrapper to solveset(..., domain=S.Reals)

smichr committed:
do not test `a == x` as `a is x` except for singletons

smichr committed:
add some very minimal simplification to inversion results

smichr committed:
update domain when solving piecewise

smichr committed:
ode: denum -> den; rewrap lines; remove inline comment

smichr committed:
test Piecewise solution for complex domain

smichr committed:
update note regarding the explicit need to give domain

smichr commented:
A strange error in mpmath has been uncovered that I will have to investigate.

asmeurer commented:
Ping @hargup and @aktech on solveset issues.

smichr commented:
- [ ] check on invert_real and invert_complex's need to use domain (b/c in complex it wasn't used)

aktech commented:
Ping me when it's ready to go.

smichr committed:
consolidate invert_real and invert_complex

Also check that the symbol is a Symbol.

smichr commented:
ping @aktech. If you can take a look, that would be appreciated.

smichr commented:
@aktech, can you give this a look?

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -714,120 +514,96 @@ def _solve_abs(f, symbol):
 
 
-def solveset(f, symbol=None, domain=S.Complexes):
+def solveset(f, symbol=None, domain=S.Complexes, check=True, _first=True, ):
```
How about keeping these flags (atleast `_first`) in the helper function? It seems like the input API of `solveset_real` is more cleaner than this.

smichr replied:
These are private flags; they could be hidden in `kwargs` but there is no need. They aren't discussed as a use option and (now) the `check` flag reads `_check`. solveset is the front door to all the routines. It's just my preference to handle the helper (for check and first) within the function rather than in "yet another helper function". If you or anyone else feels strongly about not doing so, I can refactor.

asmeurer replied:
I agree with @aktech. Part of the complexity of `solve` is the overabundance of command line flags.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -863,15 +639,14 @@ def solveset(f, symbol=None, domain=S.Complexes):
     =====
 
+    The output of `solveset` depends on the requested domain. It
+    ignores assumptions on the symbol procided by design: it will
```
procided -> provided,

smichr replied:
got it

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -863,15 +639,14 @@ def solveset(f, symbol=None, domain=S.Complexes):
 
+    The output of `solveset` depends on the requested domain. It
+    ignores assumptions on the symbol procided by design: it will
+    always be clear what domain is being used without need to
```
*always be clear about the domain being used without the need to*

smichr replied:
I modified this as

```
    The output of `solveset` depends on the requested domain rather
    than on the assumptions a symbol may have. Assumptions on symbols
    are ignored.
```

It's not really clearer since writing `x = symbol('x', real=True); solveset(foo, x)` is no more obfuscated than `dom = S.Reals; solveset(foo, x, dom)`. But the domain can be constructed from sets to be more precise than assumptions (e.g. you could set the domain to be any subset of reals).

smichr committed:
edit docstrings; make check a private flag

smichr committed:
test issue 10555

smichr commented:
It's been a while since I looked at this and having looked again, I agree with you: I will move the `_first` related material to the end of the helper.

smichr committed:
_first -> _final; move _final and _check to _solveset

smichr committed:
10430: use domain to define symbols as real in finite check

This is the one place where the symbol assumption partially
control the output. If a symbol is declared as finite, real, or
complex, those definitions are used to determine if the solution
to a Mul is the union of the solutions of the factors.

This should probably be changed so that the conditions necessary for
a solution are explicitly returned with the solution and not modified
by assumptions on the symbol.

smichr commented:
@jksuom, do you have an idea about how to handle this latest issue? I am trying to do "safe simplification" by not allowing simplify to return an expression which has lost a singularity defined by some denominator, e.g. `1/(1/x+1)` is not allowed to simplify to `x/(1 + x)`. But then this disallows `factorial(x)/gamma(x)` from simplifying to 1. I wonder if I should only disallow denominators which are **Add** instances to not disappear.

smichr commented:
If there are no more comments, @aktech , I will commit this in 12h.

aktech commented:
@smichr I am sorry, for delaying this PR. Actually I am running short of time this week and next week so can't take a thorough look write now, so I would wait for @hargup to give a thumps up (he also probably have his exams write now). Sorry for the inconvenience.

jksuom commented:
> 1/(1/x+1) is not allowed to simplify to x/(1 + x).

I'm afraid that trying to achieve this could result in problems hard to handle. In fact, there are reasons why I would rather allow this simplification.

BTW, you probably meant that ``factorial(x)/gamma(x)`` should simplify to ``x`` (for ``x`` a positive integer).

smichr commented:
@jksuom , I added the comment to the wrong PR -- I'll move it to the other one regarding Relational simplification.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -889,33 +696,49 @@ def solveset(f, symbol=None, domain=S.Complexes):
+      real domain, provide a real domain. (Using `solveset\_real`
+      does this automatically.)
 
+    >>> R = S.Reals
```
This is a useful alias. We should create a PR so that someone will be able to import it by `from sympy import I` like we can do for `I`, `E` and `pi`.

asmeurer replied:
I would be cautious about that. The existing single-letter aliases cause issues for people already, for example, `sympify('X + 1 + O')` doesn't do what many people would expect.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -25,95 +25,125 @@
+    `{-n, n}`, as the nature of `n` is unknown; rather it will return:
     `Intersection([0, oo) {n}) U Intersection((-oo, 0], {-n})`
 
+    By default, the complex domain is used but note that inverting even
```
Thanks for the docstring, it explains the differences between the real and complex domain very clearly.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -559,14 +389,13 @@ def _solve_real_trig(f, symbol):
 
 
-def _solve_as_poly(f, symbol, solveset_solver, invert_func):
+def _solve_as_poly(f, symbol, solveset_solver, invert_func, domain=S.Complexes):
```
The `solveset_solver` argument is not being used anywhere in the function body. Also it not required now, as specify the solver unambiguously using the `domain` agrument. When I originally wrote this function the `domain` argument was not present and to call the different solver I had to use the different function `solveset_real` and `solveset_complex`.

smichr replied:
removed

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -698,7 +508,7 @@ def _solve_radical(f, symbol, solveset_solver):
 
 
-def _solve_abs(f, symbol):
+def _solve_abs(f, symbol, domain=S.Complexes):
```
You should remove the default value and raise an error if the domain is complex. This is solution is valid only on the real domain. On complex domain `Abs(z) = sqrt(Re(z)**2 + Im(z)**2)` and this `Abs` is not invertible, there are uncountably many complex numbers which corresponds to particular absolute value.

smichr replied:
done and test added

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -715,117 +525,120 @@ def _solve_abs(f, symbol):
-    if not isinstance(f, (Expr, Number)):
-        raise ValueError(" %s is not a valid sympy expression" % (f))
+    # assign the solvers to use
+    solver = lambda f, x, domain=domain: _solveset(f, x, domain)
```
This is same as `solver = _solveset` also since there is not need to specify the solver for `_solve_as_poly`, you might as well do away with this line.

hargup replied:
Okay, I realized that this defaults the value of domain to the value passed in this instance of `_solveset` so this is not same as `_solveset`, anyway there is no need for line if `_solve_as_poly` stops accepting `solver` as an argument.

smichr replied:
The `solver` is needed for a variety of cases below (not only, previously, for the `poly_solver`).

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -715,117 +525,120 @@ def _solve_abs(f, symbol):
+                    elif equation.has(Abs):
+                        result += _solve_abs(f, symbol, domain)
                     else:
                         result += _solve_as_rational(equation, symbol,
```
The argument for `_solve_as_poly` also goes for `_solve_as_rational`, `domain` is enough to unambiguously specify the solvers.

smichr replied:
ok, passing of the solvers has been removed and is handled now with the domain alone.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -715,117 +525,120 @@ def _solve_abs(f, symbol):
-        return FiniteSet(*result)
-    else:
-        return result
+    if _final:
```
Can you explain this optional argument, it appears that the condition `_final=True, _check=False` doesn't make any difference in the final result. It appear that `_check` is sufficient for the purpose.

smichr replied:
Good catch -- I now only use a single flag, `_check`, which controls solution checking.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -715,117 +525,120 @@ def _solve_abs(f, symbol):
+
+            if isinstance(result, FiniteSet):
+                # check the result for invalid solutions
+                result = FiniteSet(*[s for s in result
```
We should make `domain_check` a compulsions, instead of doing it only when an optional flag is passed, because it is wrong to return a wrong result.

smichr replied:
Checking is always done for `solveset` (except during recursive calls) and that entails using `domain_check`. What may be different is that it is now only done when the final solution is ready to be returned.

hargup commented:
@smichr Thanks for this PR it greatly simplifies things. I have added my comments and after they are addressed it should be good to go.

smichr committed:
8715: remove singularities from Relational solution

smichr commented:
Let's see if tests pass after the recent edit and if there are no final comments, I will commit this. Thanks for looking it over, @hargup.

smichr committed:
review edits

smichr committed:
update trig test
